### PR TITLE
`URLClassLoader` support for `PluginFirstClassLoader`

### PR DESCRIPTION
**Pull request by basil**
#5698 revived support for `URLClassLoader` via an experimental `-Dhudson.ClassicPluginStrategy.useAntClassLoader=false` option, but this did not extend to `PluginFirstClassLoader`. This PR fills the gap by introducing a new `PluginFirstClassLoader2` based on `URLClassLoader2`, enabled with the same `-Dhudson.ClassicPluginStrategy.useAntClassLoader=false` option (i.e., off by default). The implementation delegates first to the system class loader, then to the plugin class loader, then to the parent (i.e. core) class loader. To implement `ClassLoader#getResources(String)` I used Kohsuke's `CompoundEnumeration` class, which appears to have never been used or tested before. It contained a bug, which I fixed. I also added tests for `CompoundEnumeration` which fail without the changes to `CompoundEnumeration` in `core/src/main/` and pass with them. To test `PluginFirstClassLoader2`, I loaded a few plugins with `pluginFirstClassLoader` and stepped through various scenarios in the script console with the debugger open to verify that the order of operations was as expected.

### Proposed changelog entries

* Developer: The option `-Dhudson.ClassicPluginStrategy.useAntClassLoader=false` allows experiments with plugin-first class loading alternatives.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:73]: is this the correct order for plugin-first? (I beleive this is different to the old ANTClassLoader`` which not directly consult the system classloader, but leaves it to the parent (after it has tried to load from the plugin)?

https://github.com/apache/ant/blob/002596a9bb830257250ea394a2637d57209edb3b/src/main/org/apache/tools/ant/AntClassLoader.java#L1116-L1127

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:120]: similar to previous - I think this would have been consulted as part of the parent before so should come after looking for resources in this plugin?

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:160]: ditto the enumeration order with system classloader.

File: core/src/main/java/hudson/util/CompoundEnumeration.java
Comment (jtnord) [inline core/src/main/java/hudson/util/CompoundEnumeration.java:29]: fixes an an existing bug (as cur would always be `null` and `hasMoreElements` should have thrown a `NullPointerExcetpion`-  quick search shows this was never used before.

**Review by jtnord — COMMENTED** [review]
should the system classloader be used before the plugins classloader?
It appears to be a change from how the classloader behaved before.

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:24]: would it be possible to add a test for this (there is already a [test ](https://github.com/jenkinsci/jenkins/blob/e8f2ae4e4c1ae63ae48d18f2a7b7134f317c7f7f/test/src/test/java/hudson/PluginManagerTest.java#L551-L565)for the older plugin first)

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:73]: I removed any references to the system class loader.

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:120]: I removed any references to the system class loader.

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:160]: I removed any references to the system class loader.

File: core/src/main/java/hudson/util/CompoundEnumeration.java
Comment (basil) [reply core/src/main/java/hudson/util/CompoundEnumeration.java:29]: > fixes an an existing bug (as cur would always be `null` and `hasMoreElements` should have thrown a `NullPointerExcetpion`- quick search shows this was never used before.

Yes, as I stated in the PR description:

> To implement `ClassLoader#getResources(String)` I used Kohsuke's `CompoundEnumeration` class, which appears to have never been used or tested before. It contained a bug, which I fixed.

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:23]: > there is already a [test](https://github.com/jenkinsci/jenkins/blob/e8f2ae4e4c1ae63ae48d18f2a7b7134f317c7f7f/test/src/test/java/hudson/PluginManagerTest.java#L551-L565) for the older plugin first classloader

I verified that the existing `PluginManagerTest#findResourceForPluginFirstClassLoader` passes with both `hudson.ClassicPluginStrategy.useAntClassLoader=true` (`PluginFirstClassLoader`) and `hudson.ClassicPluginStrategy.useAntClassLoader=false` (`PluginFirstClassLoader2`). Due to how early `ClassicPluginStrategy` is loaded during the tests, it was too difficult to get a test to set the flag early enough to cover both cases (even with `FlagRule`), but if and when we ever change the default value of the flag, the existing test will exercise the corresponding class loader.

Comment (jtnord) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:23]: > but if and when we ever change the default value of the flag, the existing test will exercise the corresponding class loader.

👍 

FWIW `RealJenkinsRule` may help.

**Review by jtnord — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jglick) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:21]:
```suggestion
```

Not an API, so no need to record version on which it was added.

File: core/src/main/java/hudson/PluginFirstClassLoader2.java
Comment (jglick) [inline core/src/main/java/hudson/PluginFirstClassLoader2.java:44]: Matches the implementation here but not the PR description

> The implementation delegates first to the system class loader, then to the plugin class loader, then to the parent (i.e. core) class loader.

unless I am missing something?

Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:44]: > unless I am missing something

You are missing the review by @jtnord during which I was asked to remove the delegation to the system class loader.

Comment (jtnord) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:43]: https://github.com/jenkinsci/jenkins/pull/5970#discussion_r765713610

Comment (jglick) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:43]: So the PR description is out of date and should be edited?

Comment (basil) [reply core/src/main/java/hudson/PluginFirstClassLoader2.java:43]: > So the PR description is out of date

Correct

**Review by jglick — APPROVED** [review]

**Pull request merged**
